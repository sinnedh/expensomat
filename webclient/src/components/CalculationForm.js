import React from 'react';
import {getFormFieldValue} from '../utils';

class CalculationForm extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      name: "",
      description: "",
      members: [{name: ""}],
    };

    this.addMember = this.addMember.bind(this);
    this.deleteMember = this.deleteMember.bind(this);
    this.handleMemberNameChange = this.handleMemberNameChange.bind(this);
    this.handleInputChange = this.handleInputChange.bind(this);
    this.handleSubmit = this.handleSubmit.bind(this);
  }

  addMember(event) {
    event.preventDefault();
    this.setState({"members": [...this.state.members, {name: ""}]});
  }

  deleteMember(event, index) {
    event.preventDefault();

    let members = this.state.members;
    members.splice(index, 1)

    this.setState({members});
  }

  handleSubmit(event) {
    event.preventDefault();
    this.props.handleSubmit(this.state);
  }

  handleInputChange(event) {
    const name = event.target.name;

    this.setState({
      [name]: getFormFieldValue(event.target, name)
    });
  }

  handleMemberNameChange(event, index) {
    const value = event.target.value;

    let members = this.state.members;
    members[index] = {"name": value};

    this.setState({members});
  }

  render() {
    return (
      <form>
        <label>
          Name:
          <input
            name="name"
            value={this.state.name}
            onChange={this.handleInputChange} />
        </label>
        <br />
        <label>
          Description:
          <input
            name="description"
            value={this.state.description}
            onChange={this.handleInputChange} />
        </label>
        <br />

        <h2>Members</h2>
        {this.state.members.map((member, i) =>
          <p key={i}>
            <label key={i}>
              Name:
              <input
                name={"member" + i + "-name"}
                value={member.name}
                onChange={(e) => this.handleMemberNameChange(e, i)} />
            </label>
            <input
              type="button"
              value="Delete"
              onClick={(e) => this.deleteMember(e, i)} />
          </p>
        )}

        <input type="button" value="Add member" onClick={this.addMember} />

        <input type="button" value="Create calculation" onClick={this.handleSubmit} />
      </form>
    );
  }
}

export default CalculationForm;
