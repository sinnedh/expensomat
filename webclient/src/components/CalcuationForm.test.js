import React from 'react';
import ReactDOM from 'react-dom';
import { shallow } from 'enzyme';
import CalculationForm from './CalculationForm';

it('renders', () => {
  const div = document.createElement('div');
  const onSubmit = jest.fn();
  const wrapper = shallow(<CalculationForm handleSubmit={onSubmit} />)
  expect(wrapper).toMatchSnapshot()
})

it('calls onSubmit when submit button is clicked', () => {
  const div = document.createElement('div');
  const onSubmit = jest.fn();
  const wrapper = shallow(<CalculationForm handleSubmit={onSubmit} />)
  wrapper.find('input.submit').simulate('click', { preventDefault() {} })
  expect(onSubmit).toBeCalled()
})
