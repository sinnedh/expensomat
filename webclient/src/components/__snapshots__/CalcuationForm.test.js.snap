// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <CalculationForm
    handleSubmit={[Function]}
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <label>
          Name:
          <input
                    name="name"
                    onChange={[Function]}
                    value=""
          />
</label>,
        <br />,
        <label>
          Description:
          <input
                    name="description"
                    onChange={[Function]}
                    value=""
          />
</label>,
        <br />,
        <h2>
          Members
</h2>,
        Array [
          <p>
            <label>
                        Name:
                        <input
                                    name="member0-name"
                                    onChange={[Function]}
                                    value=""
                        />
            </label>
            <input
                        onClick={[Function]}
                        type="button"
                        value="Delete"
            />
</p>,
        ],
        <input
          className="add-member"
          onClick={[Function]}
          type="button"
          value="Add member"
/>,
        <input
          className="submit"
          onClick={[Function]}
          type="button"
          value="Create calculation"
/>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            "Name:",
            <input
              name="name"
              onChange={[Function]}
              value=""
/>,
          ],
        },
        "ref": null,
        "rendered": Array [
          "Name:",
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "name": "name",
              "onChange": [Function],
              "value": "",
            },
            "ref": null,
            "rendered": null,
            "type": "input",
          },
        ],
        "type": "label",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {},
        "ref": null,
        "rendered": null,
        "type": "br",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            "Description:",
            <input
              name="description"
              onChange={[Function]}
              value=""
/>,
          ],
        },
        "ref": null,
        "rendered": Array [
          "Description:",
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "name": "description",
              "onChange": [Function],
              "value": "",
            },
            "ref": null,
            "rendered": null,
            "type": "input",
          },
        ],
        "type": "label",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {},
        "ref": null,
        "rendered": null,
        "type": "br",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Members",
        },
        "ref": null,
        "rendered": "Members",
        "type": "h2",
      },
      Object {
        "instance": null,
        "key": "0",
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <label>
              Name:
              <input
                            name="member0-name"
                            onChange={[Function]}
                            value=""
              />
</label>,
            <input
              onClick={[Function]}
              type="button"
              value="Delete"
/>,
          ],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": "0",
            "nodeType": "host",
            "props": Object {
              "children": Array [
                "Name:",
                <input
                  name="member0-name"
                  onChange={[Function]}
                  value=""
/>,
              ],
            },
            "ref": null,
            "rendered": Array [
              "Name:",
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "name": "member0-name",
                  "onChange": [Function],
                  "value": "",
                },
                "ref": null,
                "rendered": null,
                "type": "input",
              },
            ],
            "type": "label",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "onClick": [Function],
              "type": "button",
              "value": "Delete",
            },
            "ref": null,
            "rendered": null,
            "type": "input",
          },
        ],
        "type": "p",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "className": "add-member",
          "onClick": [Function],
          "type": "button",
          "value": "Add member",
        },
        "ref": null,
        "rendered": null,
        "type": "input",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "className": "submit",
          "onClick": [Function],
          "type": "button",
          "value": "Create calculation",
        },
        "ref": null,
        "rendered": null,
        "type": "input",
      },
    ],
    "type": "form",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <label>
            Name:
            <input
                        name="name"
                        onChange={[Function]}
                        value=""
            />
</label>,
          <br />,
          <label>
            Description:
            <input
                        name="description"
                        onChange={[Function]}
                        value=""
            />
</label>,
          <br />,
          <h2>
            Members
</h2>,
          Array [
            <p>
              <label>
                            Name:
                            <input
                                          name="member0-name"
                                          onChange={[Function]}
                                          value=""
                            />
              </label>
              <input
                            onClick={[Function]}
                            type="button"
                            value="Delete"
              />
</p>,
          ],
          <input
            className="add-member"
            onClick={[Function]}
            type="button"
            value="Add member"
/>,
          <input
            className="submit"
            onClick={[Function]}
            type="button"
            value="Create calculation"
/>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              "Name:",
              <input
                name="name"
                onChange={[Function]}
                value=""
/>,
            ],
          },
          "ref": null,
          "rendered": Array [
            "Name:",
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "name": "name",
                "onChange": [Function],
                "value": "",
              },
              "ref": null,
              "rendered": null,
              "type": "input",
            },
          ],
          "type": "label",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {},
          "ref": null,
          "rendered": null,
          "type": "br",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              "Description:",
              <input
                name="description"
                onChange={[Function]}
                value=""
/>,
            ],
          },
          "ref": null,
          "rendered": Array [
            "Description:",
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "name": "description",
                "onChange": [Function],
                "value": "",
              },
              "ref": null,
              "rendered": null,
              "type": "input",
            },
          ],
          "type": "label",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {},
          "ref": null,
          "rendered": null,
          "type": "br",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Members",
          },
          "ref": null,
          "rendered": "Members",
          "type": "h2",
        },
        Object {
          "instance": null,
          "key": "0",
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <label>
                Name:
                <input
                                name="member0-name"
                                onChange={[Function]}
                                value=""
                />
</label>,
              <input
                onClick={[Function]}
                type="button"
                value="Delete"
/>,
            ],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": "0",
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  "Name:",
                  <input
                    name="member0-name"
                    onChange={[Function]}
                    value=""
/>,
                ],
              },
              "ref": null,
              "rendered": Array [
                "Name:",
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "name": "member0-name",
                    "onChange": [Function],
                    "value": "",
                  },
                  "ref": null,
                  "rendered": null,
                  "type": "input",
                },
              ],
              "type": "label",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "onClick": [Function],
                "type": "button",
                "value": "Delete",
              },
              "ref": null,
              "rendered": null,
              "type": "input",
            },
          ],
          "type": "p",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "className": "add-member",
            "onClick": [Function],
            "type": "button",
            "value": "Add member",
          },
          "ref": null,
          "rendered": null,
          "type": "input",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "className": "submit",
            "onClick": [Function],
            "type": "button",
            "value": "Create calculation",
          },
          "ref": null,
          "rendered": null,
          "type": "input",
        },
      ],
      "type": "form",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
